### PR TITLE
Softforks unit

### PR DESCRIPTION
**Pull request by CodeShark**
# SoftForks Unit

This PR gathers the various soft fork mechanisms into one place making it more orderly. As a positive side effect, it will also make the VersionBits implementation simpler.

This PR depends on #6774

### CODE REVIEW

**Comment by dcousens** [comment]
concept ACK

**Comment by luke-jr** [comment]
Can you put softforks.{cpp,h} under the consensus/ directory?

**Comment by CodeShark** [comment]
@luke-jr yes.

**Comment by jonasschnelli** [comment]
Nice!
utACK.

**Comment by btcdrak** [comment]
Good, the code was all over the place before.
Concept ACK

**Comment by TheBlueMatt** [comment]
Long-term isnt the goal to just switch enforcement of most softforks to blockheight/timestamp and then use versionbits globally? (I believe most things were rolled out via a IsSuperMajority-like system, and then the code was just simplified later by replacing the checks with what-the-switchover-date-was checks).

**Comment by CodeShark** [comment]
@TheBlueMatt indeed - that's exactly what will end up happening once things are buried deeply enough.

**Comment by TheBlueMatt** [comment]
Mmm, ok, well I hadnt really read the code to begin with and I figured you were jumping the gun, but the code does simplify things on its own, so concept ack that.

**Comment by btcdrak** [comment]
I checked consensus is not affected by spinning up a new node and syncing from scratch. The node synced to the correct tip @ block 377385, `0000000000000000056d10c5f1cc9961ce2476da818a2fb4dd54ca3bcfb99a8d`

ACK

**Comment by CodeShark** [comment]
Rebased atop #6774 to make it easier to review code movements.

**Comment by jtimon** [comment]
Apart from my nits on #6774, utACK

File: src/main.cpp [hunk 1]
```diff
@@ -1710,21 +1706,9 @@ bool ConnectBlock(const CBlock& block, CValidationState& state, CBlockIndex* pin
...
-    // This logic is not necessary for memory pool transactions, as AcceptToMemoryPool
-    // already refuses previously-known transaction ids entirely.
-    // This rule was originally applied to all blocks with a timestamp after March 15, 2012, 0:00 UTC.
-    // Now that the whole chain is irreversibly beyond that time it is applied to all blocks except the
-    // two in the chain that violate it. This prevents exploiting the issue against nodes during their
-    // initial block download.
-    bool fEnforceBIP30 = (!pindex->phashBlock) || // Enforce on CreateNewBlock invocations which don't have a hash.
-                          !((pindex->nHeight==91842 && pindex->GetBlockHash() == uint256S("0x00000000000a4d0a398161ffc163c503763b1f4360639393e0e4c8e300e0caec")) ||
-                           (pindex->nHeight==91880 && pindex->GetBlockHash() == uint256S("0x00000000000743f190a18c5577a3c2d2a1f610ae9601ac046a38084ccb7cd721")));
+    // For soft fork details, look in consensus/softforks.cpp
```
Comment (jtimon) [inline src/main.cpp]: Are this comments "look in consensus/softfork.cpp" really necessary or useful?

**Comment by jtimon** [comment]
I liked '''Consensus::GetFlags''' see https://github.com/jtimon/bitcoin/commit/9ba7d0641019d7f6c08f7b15ce3a2bd2e5e8a4ac even more, but this is definitely a step in the right direction...

**Comment by dcousens** [comment]
re-ACK

**Comment by jtimon** [comment]
Wouldn't it make sense to rebase this on top of #6816 ?

File: src/consensus/softforks.h [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+enum Rule
+{
+    BIP16,
+    BIP30,
+    BIP34,
+    BIP42,
+    BIP62,
+    BIP65,
+    BIP66,
+    BIP68,
```
Comment (jtimon) [inline src/consensus/softforks.h]: In my optinion there's no need to include future softforks here (perhaps with the exception of BIP65), they can add themselves in their own PRs if they're merged after this.

Comment (dcousens) [inline src/consensus/softforks.h]: Agreed

**Comment by jtimon** [comment]
Sorry, I meant to say the following here: https://github.com/bitcoin/bitcoin/pull/6774#issuecomment-168535109

**Comment by jtimon** [comment]
rebase or close for now?

**Comment by btcdrak** [comment]
I think 7575 supersedes this.

**Comment by jtimon** [comment]
@btcdrak well, it would need to be greatly rewritten after #7575, but what I think is the general idea it's still possible, see https://github.com/jtimon/bitcoin/commit/dd446faf641350fb469865fc3586d1e088bdfb00 and https://github.com/jtimon/bitcoin/commit/032182336d84a420f33be3ab7d204c9b5f8ae8cf 
But yeah, probably a new PR would be a better idea if it's going to be rewritten.

**Comment by MarcoFalke** [comment]
Needs rebase/new pull if still relevant

**Comment by laanwj** [comment]
Closing for now

**Pull request closed without merging**